swig/lightgbmlib.i

```diff
@@ -14,18 +14,152 @@
+                                      int64_t* out_len,
+                                      double* out_result) {
+    double* data0 = (double*)jenv->GetPrimitiveArrayCritical(data, 0);
+    
+    int ret = LGBM_BoosterPredictForMatSingleRow(handle, data0, data_type, ncol, is_row_major, predict_type,
+                                     num_iteration, parameter, out_len, out_result);
+    
+    jenv->ReleasePrimitiveArrayCritical(data, data0, JNI_ABORT);
+    
+    return ret;
+  }
+  
+  int LGBM_BoosterPredictForCSRSingle(JNIEnv *jenv,
+                                     jintArray indices, 
+                                     jdoubleArray values,
+                                     int numNonZeros,
+                                     BoosterHandle handle, 
+                                     int indptr_type,
+                                     int data_type,
+                                     int64_t nelem,
+                                     int64_t num_col,
+                                     int predict_type,
+                                     int num_iteration,
+                                     const char* parameter,
+                                     int64_t* out_len,
+                                     double* out_result) {
+    // Alternatives
+    // - GetIntArrayElements: performs copy
+    // - GetDirectBufferAddress: fails on wrapped array
+    // Some words of warning for GetPrimitiveArrayCritical
+    // https://stackoverflow.com/questions/23258357/whats-the-trade-off-between-using-getprimitivearraycritical-and-getprimitivety
+    
+    jboolean isCopy;
+    int* indices0 = (int*)jenv->GetPrimitiveArrayCritical(indices, &isCopy);
+    double* values0 = (double*)jenv->GetPrimitiveArrayCritical(values, &isCopy);
+    
+    int32_t ind[2] = { 0, numNonZeros };
+    
+    int ret = LGBM_BoosterPredictForCSRSingleRow(handle, ind, indptr_type, indices0, values0, data_type, 2,
+            nelem, num_col, predict_type, num_iteration, parameter, out_len, out_result);
+    
+    jenv->ReleasePrimitiveArrayCritical(values, values0, JNI_ABORT);
+    jenv->ReleasePrimitiveArrayCritical(indices, indices0, JNI_ABORT);
+    
+    return ret;
+  }
+  
+  #include <vector>
+  #include <functional>
+  
+  struct CSRDirect {
+          jintArray indices;
+          jdoubleArray values;
+          int* indices0;
+          double* values0;
+          int size;
+  };
+  
+  int LGBM_DatasetCreateFromCSRDirect(JNIEnv *jenv,
+                                     jobjectArray arrayOfSparseVector,
+                                     int num_rows,
+                                     int64_t num_col,
+                                     const char* parameters,
+                                     const DatasetHandle reference,
+                                     DatasetHandle* out) {
+   jclass sparseVectorClass = jenv->FindClass("org/apache/spark/ml/linalg/SparseVector");
```

Comment:
seems a bit strange to have spark datastructures leak into lightgbm.  Might be good to have something like Spark in the name of the method to make it clear that this isn't even just for Java but an optimization for spark in particular.

Reply:
fair point